File: app/src/androidTest/java/com/glsebastiany/bakingapp/core/repository/DecoratedRecipesRepository.java
package com.glsebastiany.bakingapp.core.repository;

import android.content.Context;

import com.glsebastiany.bakingapp.core.SimpleIdlingResource;
import com.glsebastiany.bakingapp.repository.RecipesRepository;
import com.glsebastiany.bakingapp.repository.model.Recipe;

import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Single;

public class DecoratedRecipesRepository extends RecipesRepository {
    private SimpleIdlingResource idlingResource = new SimpleIdlingResource();

    public DecoratedRecipesRepository(Context context) {
        super(context);
    }

    public Single<List<Recipe>> getRecipes() {
        return super
                .getRecipes()
                .delay(2, TimeUnit.SECONDS) // force a delay to make use of IdlingResource necessary
                .doOnSubscribe(disposable -> idlingResource.setIdleState(false))
                .doAfterTerminate(() -> idlingResource.setIdleState(true));
    }

    public SimpleIdlingResource getIdlingResource() {
        return idlingResource;
    }
}

File: app/src/androidTest/java/com/glsebastiany/bakingapp/core/repository/DecoratedRecipesRepository.kt
package com.glsebastiany.bakingapp.core.repository

import android.content.Context
import com.glsebastiany.bakingapp.core.SimpleIdlingResource
import com.glsebastiany.bakingapp.repository.RecipesRepository
import com.glsebastiany.bakingapp.repository.model.Recipe
import io.reactivex.Single
import java.util.concurrent.TimeUnit

class DecoratedRecipesRepository(context: Context) : RecipesRepository(context) {
    val idlingResource = SimpleIdlingResource()

    override fun getRecipes(): Single<List<Recipe>> {
        return super
                .getRecipes()
                .delay(2, TimeUnit.SECONDS) // force a delay to make use of IdlingResource necessary
                .doOnSubscribe {
                    println("ois")
                    idlingResource.setIdleState(false)
                }
                .doAfterTerminate { idlingResource.setIdleState(true) }
    }
}
